File: Project2/Baitoan4/generalGraph/cpp/adjacency_matrix.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <map>

/**
 * @brief Structure to represent adjacency matrix for general graph
 */
struct AdjacencyMatrix {
    std::vector<std::vector<int>> matrixData;
    int numberOfVertices;
};

/**
 * @brief Read adjacency matrix from input file
 * @param fileName Name of input file
 * @return AdjacencyMatrix structure containing the graph data
 */
AdjacencyMatrix readAdjacencyMatrixFromFile(const std::string& fileName) {
    std::ifstream inputFile(fileName);
    if (!inputFile.is_open()) {
        std::cerr << "Error: Cannot open file " << fileName << std::endl;
        exit(1);
    }
    
    AdjacencyMatrix adjacencyMatrix;
    inputFile >> adjacencyMatrix.numberOfVertices;
    
    adjacencyMatrix.matrixData.resize(adjacencyMatrix.numberOfVertices, 
                                     std::vector<int>(adjacencyMatrix.numberOfVertices, 0));
    
    for (int rowIndex = 0; rowIndex < adjacencyMatrix.numberOfVertices; ++rowIndex) {
        for (int columnIndex = 0; columnIndex < adjacencyMatrix.numberOfVertices; ++columnIndex) {
            inputFile >> adjacencyMatrix.matrixData[rowIndex][columnIndex];
        }
    }
    
    inputFile.close();
    return adjacencyMatrix;
}

/**
 * @brief Read adjacency matrix from edge list format
 * @param fileName Name of input file containing edge list
 * @return AdjacencyMatrix structure containing the graph data
 */
AdjacencyMatrix readAdjacencyMatrixFromEdgeList(const std::string& fileName) {
    std::ifstream inputFile(fileName);
    if (!inputFile.is_open()) {
        std::cerr << "Error: Cannot open file " << fileName << std::endl;
        exit(1);
    }
    
    int numberOfVertices, numberOfEdges;
    inputFile >> numberOfVertices >> numberOfEdges;
    
    AdjacencyMatrix adjacencyMatrix;
    adjacencyMatrix.numberOfVertices = numberOfVertices;
    adjacencyMatrix.matrixData.resize(numberOfVertices, std::vector<int>(numberOfVertices, 0));
    
    for (int edgeIndex = 0; edgeIndex < numberOfEdges; ++edgeIndex) {
        int sourceVertex, targetVertex;
        inputFile >> sourceVertex >> targetVertex;
        ++adjacencyMatrix.matrixData[sourceVertex][targetVertex];
    }
    
    inputFile.close();
    return adjacencyMatrix;
}

/**
 * @brief Display adjacency matrix to console
 * @param adjacencyMatrix The adjacency matrix to display
 */
void displayAdjacencyMatrix(const AdjacencyMatrix& adjacencyMatrix) {
    std::cout << "=== Adjacency Matrix ===" << std::endl;
    std::cout << "Number of vertices: " << adjacencyMatrix.numberOfVertices << std::endl;
    
    for (int rowIndex = 0; rowIndex < adjacencyMatrix.numberOfVertices; ++rowIndex) {
        for (int columnIndex = 0; columnIndex < adjacencyMatrix.numberOfVertices; ++columnIndex) {
            std::cout << adjacencyMatrix.matrixData[rowIndex][columnIndex];
            if (columnIndex < adjacencyMatrix.numberOfVertices - 1) {
                std::cout << " ";
            }
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
}

/**
 * @brief Write adjacency matrix to output file
 * @param adjacencyMatrix The adjacency matrix to write
 * @param fileName Name of output file
 */
void writeAdjacencyMatrixToFile(const AdjacencyMatrix& adjacencyMatrix, const std::string& fileName) {
    std::ofstream outputFile(fileName);
    if (!outputFile.is_open()) {
        std::cerr << "Error: Cannot create file " << fileName << std::endl;
        return;
    }
    
    outputFile << "=== Adjacency Matrix ===" << std::endl;
    outputFile << "Number of vertices: " << adjacencyMatrix.numberOfVertices << std::endl;
    
    for (int rowIndex = 0; rowIndex < adjacencyMatrix.numberOfVertices; ++rowIndex) {
        for (int columnIndex = 0; columnIndex < adjacencyMatrix.numberOfVertices; ++columnIndex) {
            outputFile << adjacencyMatrix.matrixData[rowIndex][columnIndex];
            if (columnIndex < adjacencyMatrix.numberOfVertices - 1) {
                outputFile << " ";
            }
        }
        outputFile << std::endl;
    }
    
    outputFile.flush();
    outputFile.close();
}

/**
 * @brief Convert adjacency list to adjacency matrix
 * @param adjacencyData Adjacency list data
 * @param numberOfVertices Number of vertices in the graph
 * @return AdjacencyMatrix structure containing the converted data
 */
AdjacencyMatrix convertAdjacencyListToMatrix(const std::vector<std::vector<int>>& adjacencyData,
                                            int numberOfVertices) {
    AdjacencyMatrix adjacencyMatrix;
    adjacencyMatrix.numberOfVertices = numberOfVertices;
    adjacencyMatrix.matrixData.resize(numberOfVertices, std::vector<int>(numberOfVertices, 0));
    
    for (int sourceVertex = 0; sourceVertex < numberOfVertices; ++sourceVertex) {
        for (int targetVertex : adjacencyData[sourceVertex]) {
            ++adjacencyMatrix.matrixData[sourceVertex][targetVertex];
        }
    }
    
    return adjacencyMatrix;
}

/**
 * @brief Convert extended adjacency list to adjacency matrix
 * @param edgeInstances Edge instances from extended list
 * @param numberOfVertices Number of vertices in the graph
 * @return AdjacencyMatrix structure containing the converted data
 */
AdjacencyMatrix convertExtendedAdjacencyListToMatrix(const std::vector<std::pair<int, int>>& edgeInstances,
                                                    int numberOfVertices) {
    AdjacencyMatrix adjacencyMatrix;
    adjacencyMatrix.numberOfVertices = numberOfVertices;
    adjacencyMatrix.matrixData.resize(numberOfVertices, std::vector<int>(numberOfVertices, 0));
    
    for (const auto& edge : edgeInstances) {
        int sourceVertex = edge.first;
        int targetVertex = edge.second;
        ++adjacencyMatrix.matrixData[sourceVertex][targetVertex];
    }
    
    return adjacencyMatrix;
}

/**
 * @brief Convert adjacency map to adjacency matrix
 * @param outgoingConnections Outgoing connections from adjacency map
 * @param numberOfVertices Number of vertices in the graph
 * @return AdjacencyMatrix structure containing the converted data
 */
AdjacencyMatrix convertAdjacencyMapToMatrix(const std::map<int, std::vector<std::pair<int, std::pair<int, int>>>>& outgoingConnections,
                                           int numberOfVertices) {
    AdjacencyMatrix adjacencyMatrix;
    adjacencyMatrix.numberOfVertices = numberOfVertices;
    adjacencyMatrix.matrixData.resize(numberOfVertices, std::vector<int>(numberOfVertices, 0));
    
    for (const auto& vertexPair : outgoingConnections) {
        int sourceVertex = vertexPair.first;
        const auto& connections = vertexPair.second;
        
        for (const auto& connection : connections) {
            int targetVertex = connection.first;
            ++adjacencyMatrix.matrixData[sourceVertex][targetVertex];
        }
    }
    
    return adjacencyMatrix;
} 

File: Project2/Baitoan4/generalGraph/py/adjacency_matrix.py
"""
Adjacency Matrix implementation for General Graph
Supports self-loops and multiple edges
"""

from typing import List, Tuple
from dataclasses import dataclass


@dataclass
class AdjacencyMatrix:
    """Structure to represent adjacency matrix for general graph"""
    matrix_data: List[List[int]]
    number_of_vertices: int
    
    def __init__(self, number_of_vertices: int = 0):
        self.number_of_vertices = number_of_vertices
        self.matrix_data = [[0 for _ in range(number_of_vertices)] 
                           for _ in range(number_of_vertices)]


def read_adjacency_matrix_from_file(file_name: str) -> AdjacencyMatrix:
    """
    Read adjacency matrix from input file
    
    Args:
        file_name: Name of input file
        
    Returns:
        AdjacencyMatrix structure containing the graph data
    """
    try:
        with open(file_name, 'r') as input_file:
            number_of_vertices, number_of_edges = map(int, input_file.readline().split())
            
            adjacency_matrix = AdjacencyMatrix(number_of_vertices)
            
            for _ in range(number_of_edges):
                source_vertex, target_vertex = map(int, input_file.readline().split())
                if (0 <= source_vertex < number_of_vertices and 
                    0 <= target_vertex < number_of_vertices):
                    adjacency_matrix.matrix_data[source_vertex][target_vertex] += 1
            
            return adjacency_matrix
    except FileNotFoundError:
        print(f"Error: Cannot open file {file_name}")
        return AdjacencyMatrix()
    except Exception as e:
        print(f"Error reading file: {e}")
        return AdjacencyMatrix()


def display_adjacency_matrix(adjacency_matrix: AdjacencyMatrix) -> None:
    """
    Display adjacency matrix to console
    
    Args:
        adjacency_matrix: The adjacency matrix to display
    """
    print("=== Adjacency Matrix ===")
    print(f"Number of vertices: {adjacency_matrix.number_of_vertices}")
    
    for row in adjacency_matrix.matrix_data:
        print(" ".join(map(str, row)))
    print()


def write_adjacency_matrix_to_file(adjacency_matrix: AdjacencyMatrix, file_name: str) -> None:
    """
    Write adjacency matrix to output file
    
    Args:
        adjacency_matrix: The adjacency matrix to write
        file_name: Name of output file
    """
    try:
        with open(file_name, 'w') as output_file:
            output_file.write("=== Adjacency Matrix ===\n")
            output_file.write(f"Number of vertices: {adjacency_matrix.number_of_vertices}\n")
            
            for row in adjacency_matrix.matrix_data:
                output_file.write(" ".join(map(str, row)) + "\n")
    except Exception as e:
        print(f"Error: Cannot create file {file_name}: {e}")


def convert_adjacency_list_to_matrix(adjacency_data: List[List[int]], 
                                   number_of_vertices: int) -> AdjacencyMatrix:
    """
    Convert adjacency list to adjacency matrix
    
    Args:
        adjacency_data: The adjacency list data
        number_of_vertices: Number of vertices in the graph
        
    Returns:
        AdjacencyMatrix structure containing the converted data
    """
    adjacency_matrix = AdjacencyMatrix(number_of_vertices)
    
    for source_vertex in range(number_of_vertices):
        for target_vertex in adjacency_data[source_vertex]:
            if 0 <= target_vertex < number_of_vertices:
                adjacency_matrix.matrix_data[source_vertex][target_vertex] += 1
    
    return adjacency_matrix


def convert_extended_adjacency_list_to_matrix(edge_instances: List[Tuple[int, int]], 
                                            number_of_vertices: int) -> AdjacencyMatrix:
    """
    Convert extended adjacency list to adjacency matrix
    
    Args:
        edge_instances: List of all edge instances as (source, target) pairs
        number_of_vertices: Number of vertices in the graph
        
    Returns:
        AdjacencyMatrix structure containing the converted data
    """
    adjacency_matrix = AdjacencyMatrix(number_of_vertices)
    
    for source_vertex, target_vertex in edge_instances:
        if (0 <= source_vertex < number_of_vertices and 
            0 <= target_vertex < number_of_vertices):
            adjacency_matrix.matrix_data[source_vertex][target_vertex] += 1
    
    return adjacency_matrix


def convert_adjacency_map_to_matrix(outgoing_connections: dict, 
                                  number_of_vertices: int) -> AdjacencyMatrix:
    """
    Convert adjacency map to adjacency matrix
    
    Args:
        outgoing_connections: Map of outgoing connections
        number_of_vertices: Number of vertices in the graph
        
    Returns:
        AdjacencyMatrix structure containing the converted data
    """
    adjacency_matrix = AdjacencyMatrix(number_of_vertices)
    
    for source_vertex, connections in outgoing_connections.items():
        for target_vertex, _ in connections:
            if 0 <= target_vertex < number_of_vertices:
                adjacency_matrix.matrix_data[source_vertex][target_vertex] += 1
    
    return adjacency_matrix 